### ISSUE TITLE
`typing.get_type_hints` Can Return Incorrect Type for Annotated Metadata

### ISSUE THREAD

**Issue by cmg2146** [python/cpython#112618]
# Bug report

### Bug description:

When running the following example:
```python
from typing import Annotated, get_type_hints

class MessageCode(str):
    pass

class DisplayName(str):
    pass

class Message:
    field_a: Annotated[str, MessageCode("A")]

class Form:
    box_a: Annotated[str, DisplayName("A")]

hints = get_type_hints(Message, include_extras=True)
metadata = hints["field_a"].__metadata__[0]
print(f"MessageCode metadata type: {type(metadata)}")

hints = get_type_hints(Form, include_extras=True)
metadata = hints["box_a"].__metadata__[0]
print(f"DisplayName metadata type: {type(metadata)}")
```

the output will be:
```
MessageCode metadata type: <class '__main__.MessageCode'>
DisplayName metadata type: <class '__main__.MessageCode'>
```

when it should be:
```
MessageCode metadata type: <class '__main__.MessageCode'>
DisplayName metadata type: <class '__main__.DisplayName'>
```

This issue seems to occur only when:
* the metadata subclasses an immutable type
* when multiple instances of such metadata appear in different class definitions and they have the same value

### CPython versions tested on:

3.9

### Operating systems tested on:

Linux, Windows

### Linked PRs
* gh-112619
* gh-112628
* gh-112633

**Comment by Eclips4** [comment]
Hello!
This happening due to realization of `typing._tp_cache`(it's used to decorate the `__getitem__` method of `_SpecialForm` to reduce the number of unnecessary calls when objects are identical). Hashes of `DisplayName("A")` and `MessageCode("A")` are identical, and they're also identical (`DisplayName("A") == MessageCode("A")` evaluates to `True`).
So this example will probably help you understand the problem more deeply:
```python
from typing import Annotated, get_type_hints

class Parent:
    def __init__(self, num: int) -> None:
        self.num = num

class Foo(Parent):
    def __eq__(self, other):
        if not isinstance(other, Parent):
            return NotImplemented
        return self.num == other.num
    def __hash__(self):
        return hash(self.num)

class Bar(Parent):  # Same realization of __eq__ as in Foo but semantically it's another class :)
    def __eq__(self, other):
        if not isinstance(other, Parent):
            return NotImplemented
        return self.num == other.num
    def __hash__(self):
        return hash(self.num)

class Form:
    box_a: Annotated[str, Foo(1)]

class Message:
    field_a: Annotated[str, Bar(1)]

hints = get_type_hints(Message, include_extras=True)
metadata = hints["field_a"].__metadata__[0]
print(f"MessageCode metadata type: {type(metadata)}")

hints = get_type_hints(Form, include_extras=True)
metadata = hints["box_a"].__metadata__[0]
print(f"DisplayName metadata type: {type(metadata)}")
```
Honestly,  I don't know, should it be fixed or not, this behaviour works for years (as I see, from 2018 year).
cc @AlexWaygood @JelleZijlstra

**Comment by sobolevn** [comment]
I think that there's an easy fix for that which also keeps all tests green:

```python
>>> hints = get_type_hints(Message, include_extras=True)
>>> metadata = hints["field_a"].__metadata__[0]
>>> print(f"MessageCode metadata type: {type(metadata)}")
MessageCode metadata type: <class '__main__.MessageCode'>
>>> 
>>> hints = get_type_hints(Form, include_extras=True)
>>> metadata = hints["box_a"].__metadata__[0]
>>> print(f"DisplayName metadata type: {type(metadata)}")
DisplayName metadata type: <class '__main__.MessageCode'>
```
```                                                                                        
(.venv) ~/Desktop/cpython2  main ✗                                                        
» ./python.exe -m test test_typing
Using random seed: 342322226
0:00:00 load avg: 1.85 Run 1 test sequentially
0:00:00 load avg: 1.85 [1/1] test_typing

== Tests result: SUCCESS ==

1 test OK.

Total duration: 402 ms
Total tests: run=613 skipped=1
Total test files: run=1/1
Result: SUCCESS
```

Patch:

```
» git patch
diff --git Lib/typing.py Lib/typing.py
index b3af701f8d5..a7fa763c198 100644
--- Lib/typing.py
+++ Lib/typing.py
@@ -497,6 +497,11 @@ def __getitem__(self, parameters):
         return self._getitem(self, *parameters)
 
 
+class _AnnotatedSpecialForm(_SpecialForm, _root=True):
+    def __getitem__(self, parameters):
+        return self._getitem(self, parameters)
+
+
 class _AnyMeta(type):
     def __instancecheck__(self, obj):
         if self is Any:
@@ -2005,7 +2010,8 @@ def __mro_entries__(self, bases):
         return (self.__origin__,)
 
 
-@_SpecialForm
+@_AnnotatedSpecialForm
+@_tp_cache(typed=True)
 def Annotated(self, params):
     """Add context-specific metadata to a type.
```

I will experiment a bit with it and decide whether or not I will open a PR for it :)

**Comment by AlexWaygood** [comment]
Thanks for the bug report @cmg2146! Here's a more minimal repro:

```pycon
>>> from typing import Annotated
>>> X = Annotated[1, 1]
>>> Y = Annotated[1, True]
>>> Y
typing.Annotated[1, 1]
```

@Eclips4, I think this is definitely worth fixing :) I think `Annotated` wasn't used much for the first few years of its existence, but there's now a growing number of third-party libraries that are using it more, so it "makes sense" that this bug might have gone unnoticed until now.

**Comment by AlexWaygood** [comment]
Fixed on the `main` branch, the 3.12 branch and the 3.11 branch. The fix will be included in the next patch release for each.

Thanks again @cmg2146 for the report, and thanks @sobolevn for the fix!

**Comment by cmg2146** [comment]
> Fixed on the `main` branch, the 3.12 branch and the 3.11 branch. The fix will be included in the next patch release for each.
> 
> Thanks again @cmg2146 for the report, and thanks @sobolevn for the fix!

Thanks for fixing this! Do you know if this fix will be ported to typing_extensions?

**Comment by JelleZijlstra** [comment]
I'd accept a PR adding this to typing-extensions.
